### PR TITLE
Autocloning and pre-scanning for the dead [ready]

### PR DESCRIPTION
**Pull request by 4dplanner**
:cl:
add: pre-scanning and autocloning are back, but any records linked to living people will be marked as outdated and unusable for cloning.
/:cl:

This should make it a bit easier to clone piles of dead people after a tragedy, while not making it any easier to clone without the body. You can also scan a living person if you want to save their DNA to a diskette.

### CODE REVIEW

**Comment by AlexTheSysop42** [comment]
allow scanning living people but make it only autoclone if they were scanned while dead
autodeleting living records is dumb, if you want to keep them from being scanned just make it not let them be scanned

**Comment by 4dplanner** [comment]
It probably is worth doing both

**Comment by 4dplanner** [comment]
Closes #41808

**Comment by Time-Green** [comment]
If there's too many people needing cloning, you can link the console to an additional pod for twice the cloning

**Comment by 4dplanner** [comment]
> If there's too many people needing cloning, you can link the console to an additional pod for twice the cloning

You can, but that requires you to spend the time to actually build a pod either in advance or while the corpses are just lying there. Even then 2 pods is better than one, but still requires you to come back and scan after each batch is done and hope the ghost is responsive.

**Comment by epochayur** [comment]
>but any records linked to living people will be automatically deleted

Why?

**Comment by 4dplanner** [comment]
Same reason as why prescans were removed, I don't like bodyless cloning at the touch of a button.

**Review by optimumtact — DISMISSED** [review]

**Comment by Bombpants** [comment]
Would an upgrade, either from higher tier parts or a specific research node, allowing autocloning be more fair? Maybe require abductor tech or so super high research cost so it's not gonna be super common.

**Comment by 4dplanner** [comment]
Making it an extremely rare cool thing for the crew might work, but is not really within the scope of this PR. This is to make things less annoying for medics.

**Comment by AlexTheSysop42** [comment]
Prescan is fine. There were never any problems with prescanning. Autocloning of live scans caused the problem.

**Comment by 4dplanner** [comment]
I think this improves the system that is in place currently, though I respect that you'd prefer a full revert of the pre-scanning part.

**Comment by carshalash** [comment]
I don't think you understand what people disliked about the pr in the slightest.

**Comment by ExcessiveUseOfCobblestone** [comment]
how do you prescan a dead body lol?

File: code/game/machinery/computer/cloning.dm [hunk 1]
```diff
@@ -6,19 +6,25 @@
 	icon_screen = "dna"
 	icon_keyboard = "med_key"
 	circuit = /obj/item/circuitboard/computer/cloning
+	req_access = list(ACCESS_HEADS) //ONLY USED FOR RECORD DELETION RIGHT NOW.
 	var/obj/machinery/dna_scannernew/scanner = null //Linked scanner. For scanning.
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]:
```suggestion
	var/obj/machinery/dna_scannernew/scanner //Linked scanner. For scanning.
```

File: code/game/machinery/computer/cloning.dm [hunk 2]
```diff
@@ -6,19 +6,25 @@
...
 	circuit = /obj/item/circuitboard/computer/cloning
+	req_access = list(ACCESS_HEADS) //ONLY USED FOR RECORD DELETION RIGHT NOW.
 	var/obj/machinery/dna_scannernew/scanner = null //Linked scanner. For scanning.
 	var/list/pods //Linked cloning pods
 	var/temp = "Inactive"
 	var/scantemp_ckey
 	var/scantemp = "Ready to Scan"
+	var/menu = 1 //Which menu screen to display
+	var/list/records = list()
 	var/datum/data/record/active_record = null
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]:
```suggestion
	var/datum/data/record/active_record
```

File: code/game/machinery/computer/cloning.dm [hunk 3]
```diff
@@ -6,19 +6,25 @@
...
+	var/obj/item/disk/data/diskette = null //Mostly so the geneticist can steal everything.
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]:
```suggestion
	var/obj/item/disk/data/diskette //Mostly so the geneticist can steal everything.
```

File: code/game/machinery/computer/cloning.dm [hunk 4]
```diff
@@ -6,19 +6,25 @@
...
 	var/loading = 0 // Nice loading text
+	var/autoprocess = 0
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]: This looks like a boo from quick read

```suggestion
	var/autoprocess = FALSE
```

File: code/game/machinery/computer/cloning.dm [hunk 5]
```diff
@@ -6,19 +6,25 @@
...
 	var/obj/machinery/dna_scannernew/scanner = null //Linked scanner. For scanning.
 	var/list/pods //Linked cloning pods
 	var/temp = "Inactive"
 	var/scantemp_ckey
 	var/scantemp = "Ready to Scan"
+	var/menu = 1 //Which menu screen to display
+	var/list/records = list()
 	var/datum/data/record/active_record = null
+	var/obj/item/disk/data/diskette = null //Mostly so the geneticist can steal everything.
 	var/loading = 0 // Nice loading text
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]:
```suggestion
	var/loading = FALSE // Nice loading text
```

File: code/game/machinery/computer/cloning.dm [hunk 6]
```diff
@@ -172,7 +284,17 @@
...
+					autoprocess = TRUE
+					playsound(src, 'sound/machines/terminal_prompt_confirm.ogg', 50, 0)
+			if("stopautoprocess")
+				autoprocess = FALSE
+				playsound(src, 'sound/machines/terminal_prompt_deny.ogg', 50, 0)
+
+	else if ((href_list["scan"]) && !isnull(scanner) && scanner.is_operational())
 		scantemp = ""
 
 		loading = 1
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]:
```suggestion
		loading = TRUE
```

File: code/game/machinery/computer/cloning.dm [hunk 7]
```diff
@@ -52,6 +58,31 @@
 			else if(!. && pod.is_operational() && !(pod.occupant || pod.mess) && pod.efficiency > 5)
 				. = pod
 
+/obj/machinery/computer/cloning/process()
+	var/should_autoprocess = scanner && LAZYLEN(pods) && autoprocess
+
+	if(should_autoprocess && scanner.occupant && scanner.scan_level > 2)
+		scan_occupant(scanner.occupant)
+
+	for(var/datum/data/record/R in records)
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/cloning.dm]:
```suggestion
	for(var/the_record in records)
	    var/datum/data/record/R = the_record
```

**Review by ExcessiveUseOfCobblestone — DISMISSED** [review]
minor stuff :)

**Comment by 4dplanner** [comment]
Just to be clear, I reverted to get the autocloning code from before so that's where all the =0s are from. Might as well clear up some of them while I'm here, though.

File: code/game/machinery/computer/cloning.dm (hunk 4 above)
Comment (4dplanner) [reply code/game/machinery/computer/cloning.dm]: > a boo
Do you mean you don't like it? Bit confused

**Comment by 4dplanner** [comment]
Using that button was a mistake

**Comment by ExcessiveUseOfCobblestone** [comment]
If your issue is "prescanning" as in cloning via having the body prior rather than currently, does this actually do what you want? I'm not really sure what the endgoal is here.

**Comment by 4dplanner** [comment]
Also not sure what you mean by "as opposed to currently", given the recent changes. The original removal of pre-scanning got merged, which was what I wanted on that front, but I've been told cloning piles of bodies became really inconvenient as a side-effect. This PR is an attempt to fix that problem while retaining the original PR's removal of long-distance cloning.

**Comment by ExcessiveUseOfCobblestone** [comment]
I'm thinking 

- someone dies
- toss in cloner and save to disk before you actually start cloning (console doesn't autodelete cuz they're dead)
- they die again but this time baddie spaces them
- you realize they're not being responsive 
- pop disk and load, see if it keeps (dead) or deletes (alive)
- ; GEORGE MELONS SPACED ME KILL HIM

I mean it relies on someone else but I didn't know where you wanted to draw the line. If it deletes from disks then same thing but you put it on RO lol

"As of currently" was in relation to having the body for prescanning in the example. You had the body "before" but not "currently" yet can still "currently" clone them. Sorry for not being clear (if that explanation even helped lol).

**Comment by 4dplanner** [comment]
You're right, I didn't consider the disks properly

**Comment by 4dplanner** [comment]
I've done some more testing, and the disk only seems to give (pasted) Unique Identifier + Unique Enzymes + Structural Enzymes. This isn't enough to create a record that isn't there already, which is why I didn't think it was a problem originally. How would I create a record from a disk?

**Comment by ExcessiveUseOfCobblestone** [comment]
ah ok i never used the disk mechanic, just doing a very quick read of the code.

In that case it should be fine.

**Comment by 4dplanner** [comment]
> ah ok i never used the disk mechanic, just doing a very quick read of the code.
> 
> In that case it should be fine.

sweet, thanks for the review

**Comment by 4dplanner** [comment]
>auto-cloning sucks 
Why does autocloning of people you have the body for suck?

**Comment by Shadowflame909** [comment]
Because, not having auto-cloning makes the otherwise bored medical doctors whose job gets cleanly replaced by a medikit/toolbox gain something to periodically check throughout the round.

This would especially be useful if Pre-Scanning worked on the living though. Huh.

**Comment by 4dplanner** [comment]
I don't think clicking through a list every ten minutes pretending to be an autocloner is good enough gameplay to be worth taking into consideration.

**Comment by 4dplanner** [comment]
I'm going to change this so records are kept but marked as outdated.

**Comment by 4dplanner** [comment]
Rewriting OP to the new system

**Comment by optimumtact** [comment]
Are you ready?

**Comment by HideAndSeekLOGIC** [comment]
Ooh, some effort. +1, nice PR

**Comment by 4dplanner** [comment]
And the merge conflict arises

**Comment by ExcessiveUseOfCobblestone** [comment]
noticed a lot of `src.var` items

**Comment by 4dplanner** [comment]
> noticed a lot of `src.var` items

Gotcha, will bring those in line as well

**Comment by 4dplanner** [comment]
There's an exploit

**Comment by 4dplanner** [comment]
I'm also going to rewrite this into a nicer format using signals.

**Pull request closed without merging**